traffic_ops/traffic_ops_golang/cachegroup/cachegroups.go

```diff
@@ -92,29 +92,51 @@ func (cg *TOCacheGroup) SetID(i int) {
 	cg.ID = &i
 }
 
-// checks if a cachegroup with the given ID is in use as a parent or secondary parent.
-func isUsedByChildCache(tx *sqlx.Tx, ID int) (bool, error) {
-	pQuery := "SELECT count(*) from cachegroup WHERE parent_cachegroup_id=$1"
-	sQuery := "SELECT count(*) from cachegroup WHERE secondary_parent_cachegroup_id=$1"
-	count := 0
-
-	err := tx.QueryRow(pQuery, ID).Scan(&count)
+// Is the cachegroup being used?
+func isUsed(tx *sqlx.Tx, ID int) (bool, error) {
+
+	var servers int
+	var pid int
+	var pid2 int
+	var asns int
+
+	/* Table counts all the places you might find a single cachegroup being used.
+	 * servers | parents | secondary parents | asns
+	 * --------+---------+-------------------+------
+	 *      12 |       0 |                 2 |    0
+	 */
+	query := `
+	SELECT
```

Comment:
Generally subqueries can be considered "less efficient" than comparable joins, but we can't really make any judgements about which query is more efficient without benchmarking them. And unless this query is found to be too slow or bottlenecking the API, we shouldn't really worry about fully optimizing it anyways.

If the JOINed version is readable enough _and_ provides the correct data, I have no problem with it. But when I run this query locally I still get the wrong data:
```
to_development=# SELECT
to_development-#         COUNT(server.id) AS servers,
to_development-#         COUNT(c2.parent_cachegroup_id) AS p1,
to_development-#         COUNT(c3.secondary_parent_cachegroup_id) AS p2,
to_development-#         COUNT(asn.id) AS asns
to_development-#     FROM cachegroup AS cg
to_development-#     LEFT JOIN server ON cg.id = cachegroup
to_development-#     LEFT JOIN cachegroup AS c2 ON cg.id = c2.parent_cachegroup_id
to_development-#     LEFT JOIN cachegroup as c3 ON cg.id = c3.secondary_parent_cachegroup_id
to_development-#     LEFT JOIN asn ON cg.id = asn.cachegroup
to_development-#     WHERE cg.id = 18;
 servers |  p1  | p2 | asns
---------+------+----+------
    1720 | 1720 |  0 | 1720
```
This query is fast, readable, and gets us what we want to know (i.e. we don't actually use the counts, all we need to know is _something_ in the table is using it):
```
SELECT
    (SELECT id FROM server WHERE server.cachegroup = $1 LIMIT 1) IS NOT NULL,
    (SELECT id FROM cachegroup WHERE cachegroup.parent_cachegroup_id = $1 LIMIT 1) IS NOT NULL,
    (SELECT id FROM cachegroup WHERE cachegroup.secondary_parent_cachegroup_id = $1 LIMIT 1) IS NOT NULL,
    (SELECT id FROM asn WHERE cachegroup = $1 LIMIT 1) IS NOT NULL
```
Then we just scan the row into booleans `isUsedByFoo`, etc.

Reply:
I see the counting problem you're talking about now (previously thought it was on something else). The boolean solution looks more elegant, so I'll end up doing that. Thanks!